### DISCUSSION TITLE
AOT-compiled DLLs: how to think about resource utilization

### DISCUSSION THREAD

**Discussion by CurtHagenlocher** [dotnet/runtime#102048] [Q&A] [answered]
I'm thinking of an architecture in which multiple AOT-compiled .NET DLLs are loaded into a single process. How should I think about the per-DLL overhead of this? I assume each of these has its own garbage-collected heap. Does each then also have a dedicated finalizer thread? What about other possible sources of inefficiency? Would loading five of these be "a bit much"? What about twenty-five?

Bonus points if the overhead can be compared with the overhead of loading Go-based DLLs exposing C-compatible APIs, as Go also has automatic memory management.

(I'm aware, of course, that any shared source dependencies would get individual copies in the AOT-compiled DLLs. In some respects, this is a feature and not a bug.)

**Comment by MichalStrehovsky** [comment, marked as answer]
Yes, each would have their own GC heap, with their own threads for doing background garbage collection, with their own finalizer thread.

I think 5 would still be okay. 25 would probably start to be too much. There are certain behaviors that don't scale well. For example, the GC will by default try to reserve 256 GB of address space on 64-bit systems just to make sure it has continuous block of memory for everything. One cannot reserve this amount many times. The amount can be adjusted by setting heap hard limit but setting it too low might start to interfere with perf or even functionality (OutOfMemoryException).

> Bonus points if the overhead can be compared with the overhead of loading Go-based DLLs exposing C-compatible APIs, as Go also has automatic memory management.

I don't think Go actually supports loading multiple DLLs with their own separate Go runtime. [Here](https://stackoverflow.com/questions/71314253/how-to-use-golang-c-shared-library-in-go) is someone trying to load a Go DLL into a Go app and failing and [here](https://github.com/golang/go/issues/50304#issuecomment-999302888) is someone trying to load a Go DLL from a C++ DLL from a Go app with a Go maintainer saying one can't have two Go runtimes in the same process. Go doesn't play well with anything within the same process, including itself - I was involved in a bug report in the past where I found out [exception handling stops working properly](https://github.com/dotnet/runtime/issues/92687#issuecomment-1738542563) in the process once Go runtime initializes in it.

**Reply by CurtHagenlocher** [reply]
Thanks! That's quite a sobering thread.

**Reply by MichalStrehovsky** [reply]
FWIW, it would not be completely unthinkable to support a mode where all the DLLs that are compiled with _compatible version_ of the runtime use a single runtime instance (with a single GC heap). "Too many people are building plugins with .NET" is not a current problem so we have not invested in solving it. But it seems solvable. "Too many people using .NET AOT" is a good problem to have...

**Reply by CurtHagenlocher** [reply]
Sure; it's a bit silly to invest in a use-case before there's a demonstrated need. The space I live in (database connectivity) is one of the few places where there are both well-defined contracts that would be amenable to something like this and where people regularly consume binary drivers -- often proprietary closed-source ones.

Referencing thread: https://github.com/apache/arrow-adbc/issues/1841

**Reply by MikeHillberg** [reply]
An interesting thing I see is that I can debug only one of the .Nets in the process. (My configuration is slightly different: an AOT app loading an AOT DLL.) I can debug the app but can't find a way to debug the DLL. E.g. I can debug the app or attach to it, but when I break into the debugger, the DLL threads keep running.

**Reply by MichalStrehovsky** [reply]
> An interesting thing I see is that I can debug only one of the .Nets in the process. (My configuration is slightly different: an AOT app loading an AOT DLL.) I can debug the app but can't find a way to debug the DLL. E.g. I can debug the app or attach to it, but when I break into the debugger, the DLL threads keep running.

How do you debug? If this is Visual Studio, make sure you attach using the native debug engine. Managed (or mixed mode) debugger gets really confused with aot binaries (and source breakpoints).

**Reply by MikeHillberg** [reply]
Good idea, yes, I was attaching with the managed debugger. Attaching with mixed mode works though; I can set a bp in code running in either CLR instance

**Comment by neon-sunset** [comment]
In addition to the answer provided by Michal, you may want to ensure the binaries use WKS GC and have limits reduced as mentioned (at the risk of OOM however, but worth a try). This way the lowest count of threads per .dll will be just 3 and their base memory footprint can also be lower (though, as tested, WKS GC, under light but constant allocation, ends up taking more memory than SRV GC + DATAS).

You likely can achieve better scalability with trial and error, but at the point of having 20 DLLs - might as well investigate using JIT and loading assemblies normally, which would be much more efficient (more efficient and scalable than Go too, at the cost of not being an isolated graph of dependencies).

**Reply by CurtHagenlocher** [reply]
Part of the goal of this would have been to load code built against .NET 8+ into an application built for .NET 4.7. The reason for AOT in this case would be to avoid having to install/distribute the .NET 8+ runtime while still being able to execute code that requires newer .NET features. If it were an option to distribute the newer runtime, we'd instead have been looking to port the application.

**Reply by josephmoresena** [reply]
In that scenario, I don't know if the hostfx library could be used directly? I call it a mini JNI for .NET... I don't think it should conflict with the .NET Framework because they're actually different environments.

@MichalStrehovsky Would that be better or worse than having AOT libraries?

**Reply by MichalStrehovsky** [reply]
> @MichalStrehovsky Would that be better or worse than having AOT libraries?

I'd check if debugging works first and foremost. The main problem with multiple **JIT-based** .NET runtimes being loaded in the process is that you may only be able to debug one of them and you won't get even as much as a stack and locals for the other one.

**Comment by buybackoff** [comment]
I have searched a lot over the web to find a definitive answer if AOTed library can be hosted by JITed .NET host. This thread is the closest.

If two AOTed libraries can co-exist in a host, it means there are no symbol conflicts between AOTed ones. By extension, if the JITed .NET host had any symbol that conflicts with AOTed library, that would mean the symbol name is fixed and two AOTed libraries would not be able to co-exist.

But is this "officially" supported and stable use case? In my case, I have a very latency sensitive loop that only requires basic setup and can be managed by volatile memory. That loop should be isolated from parent GC pauses, which happen a lot. I could do a subprocess with arguments + small shared memory file, but compiling to AOT and PInvoke-ing it seems simpler.

**Reply by MichalStrehovsky** [reply]
We don't have any tests in the dotnet/runtime repo that I know of, but I do know of a couple examples where people built a [.NET host](https://learn.microsoft.com/en-us/dotnet/core/tutorials/netcore-hosting) using native AOT and that results in the same thing (JIT-based and AOT-based .NET coexisting in the same process).

**Reply by josephmoresena** [reply]
I’ve done several proof-of-concept tests on this, and indeed, the answer is that it is possible, viable, and stable. The problem is that libraries compiled with NativeAOT cannot be unloaded, so in a hot-reload scenario, other strategies must be considered.

In a similar requirement at the company I work for, a host was needed that allowed the automatic update of the native resource (DLL), but Native AOT libraries are not unloadable and also cannot be loaded into memory using tools like 'MemoryModule'. Unfortunately, the main process for this case could not be restarted or terminated on demand.

Finally what happened was that different versions of the native resource remained loaded in the same process, but the main process only kept the latest version active. However, this led to memory usage and, at some point, CPU consumption, since each version had its own independent GC activity—even if the obsolete versions were no longer accessed, the GC within those libraries remained alive.
